File: backend/src/controllers/turnosController.js
import { pool } from '../config/database.js';

// Obtener todos los turnos disponibles
export const obtenerTurnos = async (req, res) => {
  try {
    const [turnos] = await pool.query(
      'SELECT id_turno, hora_inicio, hora_fin FROM turno ORDER BY hora_inicio'
    );

    res.json({
      success: true,
      turnos
    });

  } catch (error) {
    console.error('Error al obtener turnos:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

// Verificar disponibilidad de un turno específico
export const verificarDisponibilidadTurno = async (req, res) => {
  try {
    const { id } = req.params;
    const { fecha, sala, edificio } = req.query;

    if (!fecha) {
      return res.status(400).json({
        error: 'Fecha es requerida'
      });
    }

    // Verificar que el turno exista
    const [turnoRows] = await pool.query(
      'SELECT * FROM turno WHERE id_turno = ?',
      [id]
    );

    if (turnoRows.length === 0) {
      return res.status(404).json({
        error: 'Turno no encontrado'
      });
    }

    const turno = turnoRows[0];

    // Si se especifica sala, verificar solo esa sala
    if (sala && edificio) {
      const [reservas] = await pool.query(
        `SELECT id_reserva FROM reserva
         WHERE nombre_sala = ? AND edificio = ? AND fecha = ? AND id_turno = ?
         AND estado IN ('activa', 'finalizada')`,
        [sala, edificio, fecha, id]
      );

      return res.json({
        success: true,
        id_turno: turno.id_turno,
        hora_inicio: turno.hora_inicio,
        hora_fin: turno.hora_fin,
        fecha,
        sala,
        edificio,
        disponible: reservas.length === 0
      });
    }

    // Si no se especifica sala, obtener todas las salas disponibles
    const [todasSalas] = await pool.query(
      'SELECT nombre_sala, edificio FROM sala ORDER BY edificio, nombre_sala'
    );

    const [reservas] = await pool.query(
      `SELECT nombre_sala, edificio FROM reserva
       WHERE fecha = ? AND id_turno = ?
       AND estado IN ('activa', 'finalizada')`,
      [fecha, id]
    );

    const salasReservadas = new Set(
      reservas.map(r => `${r.nombre_sala}|${r.edificio}`)
    );

    const salasDisponibles = todasSalas.filter(
      s => !salasReservadas.has(`${s.nombre_sala}|${s.edificio}`)
    );

    res.json({
      success: true,
      id_turno: turno.id_turno,
      hora_inicio: turno.hora_inicio,
      hora_fin: turno.hora_fin,
      fecha,
      disponible: salasDisponibles.length > 0,
      total_salas: todasSalas.length,
      salas_disponibles: salasDisponibles
    });

  } catch (error) {
    console.error('Error al verificar disponibilidad del turno:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

// Obtener turnos disponibles para una fecha y sala específica
export const obtenerTurnosDisponibles = async (req, res) => {
  try {
    const { fecha, sala, edificio } = req.query;

    if (!fecha || !sala || !edificio) {
      return res.status(400).json({
        error: 'Fecha, sala y edificio son requeridos'
      });
    }

    // Obtener todos los turnos
    const [turnos] = await pool.query(
      'SELECT id_turno, hora_inicio, hora_fin FROM turno ORDER BY hora_inicio'
    );

    // Obtener reservas existentes
    const [reservas] = await pool.query(
      `SELECT id_turno FROM reserva
       WHERE nombre_sala = ? AND edificio = ? AND fecha = ?
       AND estado IN ('activa', 'finalizada')`,
      [sala, edificio, fecha]
    );

    const turnosReservados = new Set(reservas.map(r => r.id_turno));

    const turnosDisponibles = turnos.filter(
      t => !turnosReservados.has(t.id_turno)
    );

    res.json({
      success: true,
      fecha,
      sala,
      edificio,
      turnos_disponibles: turnosDisponibles,
      total_turnos: turnos.length,
      turnos_ocupados: turnosReservados.size
    });

  } catch (error) {
    console.error('Error al obtener turnos disponibles:', error);
    res.status(500).json({
      error: 'Error en el servidor'
    });
  }
};

File: backendP/src/controllers/turnos_controller.py
from fastapi import HTTPException
from src.config.database import get_connection

async def obtener_turnos():
    """Obtener todos los turnos disponibles"""
    try:
        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            'SELECT id_turno, hora_inicio, hora_fin FROM turno ORDER BY hora_inicio'
        )
        turnos = cursor.fetchall()

        # Convertir time objects a strings
        for turno in turnos:
            turno['hora_inicio'] = str(turno['hora_inicio'])
            turno['hora_fin'] = str(turno['hora_fin'])

        cursor.close()
        conn.close()

        return {
            "success": True,
            "turnos": turnos
        }

    except Exception as error:
        print(f'Error al obtener turnos: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def verificar_disponibilidad_turno(id: int, fecha: str, sala: str = None, edificio: str = None):
    """Verificar disponibilidad de un turno específico"""
    try:
        if not fecha:
            raise HTTPException(
                status_code=400,
                detail="Fecha es requerida"
            )

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        # Verificar que el turno exista
        cursor.execute('SELECT * FROM turno WHERE id_turno = %s', (id,))
        turno_rows = cursor.fetchall()

        if len(turno_rows) == 0:
            cursor.close()
            conn.close()
            raise HTTPException(
                status_code=404,
                detail="Turno no encontrado"
            )

        turno = turno_rows[0]

        # Si se especifica sala, verificar solo esa sala
        if sala and edificio:
            cursor.execute(
                """SELECT id_reserva FROM reserva
                WHERE nombre_sala = %s AND edificio = %s AND fecha = %s AND id_turno = %s
                AND estado IN ('activa', 'finalizada')""",
                (sala, edificio, fecha, id)
            )
            reservas = cursor.fetchall()

            cursor.close()
            conn.close()

            return {
                "success": True,
                "id_turno": turno['id_turno'],
                "hora_inicio": str(turno['hora_inicio']),
                "hora_fin": str(turno['hora_fin']),
                "fecha": fecha,
                "sala": sala,
                "edificio": edificio,
                "disponible": len(reservas) == 0
            }

        # Si no se especifica sala, obtener todas las salas disponibles
        cursor.execute(
            'SELECT nombre_sala, edificio FROM sala ORDER BY edificio, nombre_sala'
        )
        todas_salas = cursor.fetchall()

        cursor.execute(
            """SELECT nombre_sala, edificio FROM reserva
            WHERE fecha = %s AND id_turno = %s
            AND estado IN ('activa', 'finalizada')""",
            (fecha, id)
        )
        reservas = cursor.fetchall()

        salas_reservadas = {f"{r['nombre_sala']}|{r['edificio']}" for r in reservas}

        salas_disponibles = [
            s for s in todas_salas
            if f"{s['nombre_sala']}|{s['edificio']}" not in salas_reservadas
        ]

        cursor.close()
        conn.close()

        return {
            "success": True,
            "id_turno": turno['id_turno'],
            "hora_inicio": str(turno['hora_inicio']),
            "hora_fin": str(turno['hora_fin']),
            "fecha": fecha,
            "disponible": len(salas_disponibles) > 0,
            "total_salas": len(todas_salas),
            "salas_disponibles": salas_disponibles
        }

    except HTTPException:
        raise
    except Exception as error:
        print(f'Error al verificar disponibilidad del turno: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )


async def obtener_turnos_disponibles(fecha: str, sala: str, edificio: str):
    """Obtener turnos disponibles para una fecha y sala específica"""
    try:
        if not fecha or not sala or not edificio:
            raise HTTPException(
                status_code=400,
                detail="Fecha, sala y edificio son requeridos"
            )

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        # Obtener todos los turnos
        cursor.execute(
            'SELECT id_turno, hora_inicio, hora_fin FROM turno ORDER BY hora_inicio'
        )
        turnos = cursor.fetchall()

        # Obtener reservas existentes
        cursor.execute(
            """SELECT id_turno FROM reserva
            WHERE nombre_sala = %s AND edificio = %s AND fecha = %s
            AND estado IN ('activa', 'finalizada')""",
            (sala, edificio, fecha)
        )
        reservas = cursor.fetchall()

        turnos_reservados = {r['id_turno'] for r in reservas}

        turnos_disponibles = [
            {
                "id_turno": t['id_turno'],
                "hora_inicio": str(t['hora_inicio']),
                "hora_fin": str(t['hora_fin'])
            }
            for t in turnos
            if t['id_turno'] not in turnos_reservados
        ]

        # Convertir time objects a strings en todos los turnos
        for turno in turnos:
            turno['hora_inicio'] = str(turno['hora_inicio'])
            turno['hora_fin'] = str(turno['hora_fin'])

        cursor.close()
        conn.close()

        return {
            "success": True,
            "fecha": fecha,
            "sala": sala,
            "edificio": edificio,
            "turnos_disponibles": turnos_disponibles,
            "total_turnos": len(turnos),
            "turnos_ocupados": len(turnos_reservados)
        }

    except HTTPException:
        raise
    except Exception as error:
        print(f'Error al obtener turnos disponibles: {error}')
        raise HTTPException(
            status_code=500,
            detail="Error en el servidor"
        )

